File: sdks/canary/python/scenarios/d_session_get.py
#!/usr/bin/env python3
# Copyright (C) 2026 Michael Kao
# SPDX-License-Identifier: AGPL-3.0-or-later
"""D-SESSION-GET canary — Python SDK"""

from __future__ import annotations

import sys
import os

sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), ".."))

from canary import (
    Runner,
    Config,
    config_from_env,
    wait_active,
    ensure_session_with_retry,
    raw_do,
)
from llmsafespaces import LLMSafeSpaces


def run(r: Runner, cfg: Config) -> None:
    c = LLMSafeSpaces(cfg.api_url, api_key=cfg.api_key, timeout=60.0)
    ws_id = None
    try:
        ok, ws = r.assert_no_error(
            lambda: c.workspaces.create(
                name="canary-py-sess-get", runtime="base", storage_size="1Gi"
            ),
            "create: no error",
        )
        if not ok:
            return
        ws_id = ws.id

        phase = wait_active(c, ws_id)
        r.assert_(phase == "Active", "reach-active", f"got {phase!r}")
        if phase != "Active":
            return

        try:
            sess = ensure_session_with_retry(c, ws_id, 5)
        except Exception as e:
            r.fail("ensure-session: no error", str(e))
            return
        r.ok("ensure-session: no error")
        sid = sess.sessionId

        ok2, sess_obj = r.assert_no_error(
            lambda: c.sessions.get(ws_id, sid), "get-session: no error"
        )
        if ok2 and sess_obj is not None:
            r.assert_("id" in sess_obj, "get-session: has id")
            r.assert_("title" in sess_obj, "get-session: has title")

        r.assert_no_error(
            lambda: c.sessions.rename(ws_id, sid, "canary-py-renamed"),
            "rename: no error",
        )

        ok3, sess_after = r.assert_no_error(
            lambda: c.sessions.get(ws_id, sid), "get-after-rename: no error"
        )
        if ok3 and sess_after is not None:
            r.assert_(
                sess_after.get("title") == "canary-py-renamed",
                "get-after-rename: title updated",
                repr(sess_after.get("title")),
            )

        r.assert_error(
            lambda: c.sessions.get(ws_id, "ses_nonexistent000000"),
            "get-nonexistent-session: error",
        )

        s, _ = raw_do(
            "GET",
            f"{cfg.api_url}/api/v1/workspaces/{ws_id}/sessions/..%2Fetc",
            cfg.api_key,
        )
        r.assert_(s in (400, 404), "path-traversal-session: 400 or 404", str(s))

    finally:
        if ws_id:
            try:
                c.workspaces.delete(ws_id)
            except Exception:
                pass


if __name__ == "__main__":
    r = Runner("session-get")
    cfg = config_from_env()
    run(r, cfg)
    r.print()
    sys.exit(r.exit_code())

File: sdks/canary/typescript/scenarios/d-session-get.ts
// Copyright (C) 2026 Michael Kao
// SPDX-License-Identifier: AGPL-3.0-or-later
// D-SESSION-GET canary — TypeScript SDK

import { LLMSafeSpaces } from '../../../typescript/src/index.js';
import { Runner, Config, configFromEnv, nodeFetch, rawDo, waitActive, ensureSessionWithRetry } from '../canary.js';

async function run(r: Runner, cfg: Config): Promise<void> {
  const c = new LLMSafeSpaces({ baseUrl: cfg.apiUrl, apiKey: cfg.apiKey, timeout: 60000, fetch: nodeFetch as any });
  let wsId: string | null = null;
  try {
    const [ok, ws] = await r.assertNoError(
      () => c.workspaces.create({ name: 'canary-ts-sess-get', runtime: 'base', storageSize: '1Gi' }),
      'create: no error');
    if (!ok || !ws) return;
    wsId = ws.id;

    const phase = await waitActive(c, wsId);
    r.assert(phase === 'Active', 'reach-active', `got "${phase}"`);
    if (phase !== 'Active') return;

    const sess = await ensureSessionWithRetry(c, wsId, 5);
    r.ok('ensure-session: no error');
    const sid = sess.sessionId;

    const [ok2, sessData] = await r.assertNoError(
      () => c.sessions.get(wsId!, sid), 'get-session: no error');
    if (ok2 && sessData) {
      r.assert((sessData as any).id === sid, 'get-session: id matches', `${(sessData as any).id} vs ${sid}`);
      r.assert(typeof (sessData as any).title === 'string', 'get-session: has title field');
    }

    await r.assertNoError(
      () => c.sessions.rename(wsId!, sid, 'canary-renamed'), 'rename-session: no error');

    const [ok3, sessData2] = await r.assertNoError(
      () => c.sessions.get(wsId!, sid), 'get-session-after-rename: no error');
    if (ok3 && sessData2) {
      r.assert((sessData2 as any).title === 'canary-renamed', 'get-session: title updated',
        `got "${(sessData2 as any).title}"`);
    }

    await r.assertError(
      () => c.sessions.get(wsId!, 'ses_nonexistent000000'),
      'get-nonexistent-session: error');

    const [s] = await rawDo('GET',
      `${cfg.apiUrl}/api/v1/workspaces/${wsId}/sessions/..%2F..%2Fetc%2Fpasswd`,
      cfg.apiKey);
    r.assert(s === 400, 'path-traversal: 400', `got ${s}`);

  } finally {
    if (wsId) { try { await c.workspaces.delete(wsId); } catch {} }
  }
}

async function main() {
  const r = new Runner('session-get');
  await run(r, configFromEnv());
  r.print();
  process.exit(r.exitCode());
}
main().catch(e => { console.error(e); process.exit(1); });
